File: src/javascript/algorithms/array/32.oddOccurrences/oddOccurrences.js
// Optimal solution using XOR: TC: O(n), SC: O(1)
function findOddNumber(nums) {
    let result = null;

    for(const num of nums) {
        result ^= num;
    }

    return result;
}

// Using Object/Map with frequencies: TC: O(n), SC: O(n)
function findOddNumberObj(nums) {
    const freqObj = {};

    for(const num of nums) {
        freqObj[num] = (freqObj[num] || 0) + 1;
    }

    for(const [key, value] of Object.entries(freqObj)) {
        if(value%2 !== 0) {
            return parseInt(key);
        }
    }

    return null;
}

// Using Object keys : TC: O(n), SC: O(n)
function findOddNumberOccurrence(nums) {
    const obj = {};
    for(const num of nums) {
        if(obj[num]) {
            delete obj[num];
        } else {
            obj[num] = true;
        }
    }

    return Number(Object.keys(obj)[0]);
}

const nums1 = [9, 3, 9, 3, 9, 7, 9];
const nums2 = [1, 1, 2];
const nums3 = [4];
const nums4 = [5, 5, 5];


console.log(findOddNumber(nums1)); // 7
console.log(findOddNumber(nums2)); // 2
console.log(findOddNumber(nums3)); // 4
console.log(findOddNumber(nums4)); // 5


console.log(findOddNumberObj(nums1)); // 7
console.log(findOddNumberObj(nums2)); // 2
console.log(findOddNumberObj(nums3)); // 4
console.log(findOddNumberObj(nums4)); // 5

console.log(findOddNumberOccurrence(nums1)); // 7
console.log(findOddNumberOccurrence(nums2)); // 2
console.log(findOddNumberOccurrence(nums3)); // 4
console.log(findOddNumberOccurrence(nums4)); // 5

File: src/python/algorithms/array/32.oddOccurrences/odd_occurrences.py
"""
Odd Occurrences in Array

Given an array of integers, find the element that appears an odd number of times.
There is exactly one such element.

Time Complexity: O(n)
Space Complexity: O(1)
"""


def find_odd_number(nums):
    """
    Using XOR - optimal solution.
    TC: O(n), SC: O(1)
    """
    result = 0
    
    for num in nums:
        result ^= num
    
    return result


def find_odd_number_dict(nums):
    """
    Using dictionary for frequencies.
    TC: O(n), SC: O(n)
    """
    freq = {}
    
    for num in nums:
        freq[num] = freq.get(num, 0) + 1
    
    for key, value in freq.items():
        if value % 2 != 0:
            return key
    
    return None


def find_odd_number_toggle(nums):
    """
    Using dictionary with toggle (delete if exists).
    TC: O(n), SC: O(n)
    """
    obj = {}
    
    for num in nums:
        if num in obj:
            del obj[num]
        else:
            obj[num] = True
    
    return list(obj.keys())[0] if obj else None


# Test cases
if __name__ == "__main__":
    test_cases = [
        [9, 3, 9, 3, 9, 7, 9],   # Output: 7
        [1, 2, 1, 2, 3],         # Output: 3
        [1],                      # Output: 1
        [5, 5, 5],               # Output: 5
    ]
    
    for nums in test_cases:
        result1 = find_odd_number(nums)
        result2 = find_odd_number_dict(nums)
        result3 = find_odd_number_toggle(nums)
        print(f"Input: {nums}")
        print(f"  XOR: {result1}, Dict: {result2}, Toggle: {result3}")
